LuaRules/Gadgets/unit_temporaryarmor.lua

```diff
@@ -0,0 +1,170 @@
+if not gadgetHandler:IsSyncedCode() then -- SYNCED
+	return
+end
+
+function gadget:GetInfo()
+	return {
+		name      = "Armor States",
+		desc      = "Haxy Armors",
+		author    = "Shaman",
+		date      = "02-28-2023",
+		license   = "CC-0",
+		layer     = 0,
+		enabled   = true,
+	}
+end
+
+local IterableMap = VFS.Include("LuaRules/Gadgets/Include/IterableMap.lua")
+local handledUnits = IterableMap.New() -- which units currently have armor. Table structure: {duration = number, armorValue = number}
+local armoredUnits = {} -- table that holds units with armor.
+local watchWeapons = {} -- table that tells PreDamageWantedDefs which weapondefs we want
+local bufferProjectiles = {} -- stores information we may need for ranged units.
+local deleteProjectiles = {} -- deletes from the buffer.
+
+-- configs --
+local MAXARMOR = 0.1
+local CHECKTIME = 3 -- 10 hz or about 100ms
+
+-- Speed ups --
+-- These just make the code run faster.
+
+local spValidUnitID = Spring.ValidUnitID
+local spSetUnitArmored = Spring.SetUnitArmored
+local INLOS = {inlos = true}
+
+-- Config --
+local configs = {}
+for i = 1, #WeaponDefs do -- Iterate through every weapon def. WeaponDefs is an ordered table (meaning we have entries between 1 and some arbitrary number [like 863]
+	local weaponDef = WeaponDefs[i] -- Store the current metatable so we can reference it.
+	if weaponDef.customParams.grants_armor then -- does this weapon grant armor?
+		local armorValue = tonumber(weaponDef.customParams.grants_armor)
```

Comment:
use grants_armor custom param field (needs to be between 1 and 0). This is equal to 100% - ValueYouWantToReduce.

Example:
You want to reduce damage by 10%, set this to be 0.9.